lib/mongo_client.js

```diff
@@ -241,28 +243,35 @@ MongoClient.prototype.close = function(force, callback) {
     force = false;
   }
 
-  const client = this;
   return maybePromise(this, callback, cb => {
     const completeClose = err => {
-      client.emit('close', client);
+      this.emit('close', this);
 
-      if (!(client.topology instanceof NativeTopology)) {
-        for (const item of client.s.dbCache) {
-          item[1].emit('close', client);
+      if (!(this.topology instanceof NativeTopology)) {
+        for (const item of this.s.dbCache) {
+          item[1].emit('close', this);
         }
       }
 
-      client.removeAllListeners('close');
+      this.removeAllListeners('close');
+
+      // clear out references to old topology
+      this.topology = undefined;
+      this.s.dbCache = new Map();
+
+      // restore options
+      this.s.options = Object.assign({}, this.originalOptions);
```

Comment:
`this.s.options` can get mutated by `connect`, so we need to restore it to what it was originally constructed with. Otherwise we get the following warnings, as reported in one of the issues:

```
the options [servers] is not supported
the options [caseTranslate] is not supported
the options [dbName] is not supported
```